retrieval/raw/raw_endpoint.go

```diff
@@ -114,13 +124,145 @@ func downloadToFile(c *Client, method, sourceName string, workDir *os.File, URL
+	var to = getType(filename)
+	out := reflect.New(reflect.TypeOf(to))
+	err := json.NewDecoder(reader).Decode(out.Interface())
+
+	if err != nil {
+		return err
+	}
+	to = sanitizeData(out.Elem().Interface())
+
+	data, err := json.Marshal(to)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(writer, bytes.NewReader(data))
+	return err
+}
+
+func getType(filename string) interface{} {
+	var to interface{}
+	switch filename {
+	case Nodes:
+		to = NodeList{}
+	case Namespaces:
+		to = NamespaceList{}
+	case Pods:
+		to = PodList{}
+	case PersistentVolumes:
+		to = PersistentVolumeList{}
+	case PersistentVolumeClaims:
+		to = PersistentVolumeClaimList{}
+	case AgentMeasurement:
+		to = CldyAgent{}
+	case Services, ReplicationControllers:
+		to = LabelMapMatchedResourceList{}
+	case Deployments, ReplicaSets, Jobs, DaemonSets:
+		to = LabelSelectorMatchedResourceList{}
+	}
+	return to
+}
+
+func sanitizeData(to interface{}) interface{} {
+	switch to.(type) {
+	case LabelSelectorMatchedResourceList:
+		return sanitizeSelectorMatchedResourceList(to)
+	case PodList:
+		return sanitizePodList(to)
+	case LabelMapMatchedResourceList:
+		return sanitizeMapMatchedResourceList(to)
+	case NamespaceList:
+		return sanitizeNamespaceData(to)
+	}
+	return to
+}
+
+func sanitizeNamespaceData(to interface{}) interface{} {
+	cast := to.(NamespaceList)
+	for i, item := range cast.Items {
+		item.ObjectMeta.ManagedFields = nil
+		cast.Items[i] = item
+	}
+	return cast
+}
+
+func sanitizeSelectorMatchedResourceList(to interface{}) interface{} {
+	cast := to.(LabelSelectorMatchedResourceList)
+	for i, item := range cast.Items {
+
+		// stripping env var and related data from the object
+		item.ObjectMeta.ManagedFields = nil
+		if _, ok := item.ObjectMeta.Annotations[KubernetesLastAppliedConfig]; ok {
+			delete(item.ObjectMeta.Annotations, KubernetesLastAppliedConfig)
+		}
+		cast.Items[i] = item
+	}
+	return cast
+}
+
+func sanitizePodList(to interface{}) interface{} {
+	cast := to.(PodList)
+	for i, item := range cast.Items {
+
+		// stripping env var and related data from the object
+		item.ObjectMeta.ManagedFields = nil
+		if _, ok := item.ObjectMeta.Annotations[KubernetesLastAppliedConfig]; ok {
+			delete(item.ObjectMeta.Annotations, KubernetesLastAppliedConfig)
+		}
+		for j, container := range item.Spec.Containers {
+			item.Spec.Containers[j] = sanitizeContainer(container)
+		}
+		for j, container := range item.Spec.InitContainers {
+			item.Spec.InitContainers[j] = sanitizeContainer(container)
+		}
+		cast.Items[i] = item
+	}
+	return cast
+}
+
+func sanitizeContainer(container v1.Container) v1.Container {
+	container.Env = nil
+	container.Command = nil
+	container.Args = nil
+	container.ImagePullPolicy = ""
+	container.LivenessProbe = nil
+	container.StartupProbe = nil
+	container.ReadinessProbe = nil
+	container.TerminationMessagePath = ""
+	container.TerminationMessagePolicy = ""
+	container.SecurityContext = nil
+	return container
+}
```

Comment:
It makes me concern that we start a risk pattern to manually set a lot fields to nil or "". TBH, I didn't verify whether the logic is correct or not unless I cross-check denor logic which is time consuming. If you use any methodology to get these fields, different logic in different files, etc, I am glad to hear.
Here some risks in my head.
1. It is low readability, for example, the reader will always ask why do we need to set  ManagedFields to nil. (I guess all these field are not used by our backend, but that is not clear from the code)
2. It is risk if k8s data type is updated and how do we guarantee the backward compatible. For example, if one future version removes/rename the field container.SecurityContext etc
3. It is difficult to get what data we need/collect.. We could only tell what data we don't need and change to empty.  It also indicates this logic is difficult to test.

I remember we discuss define our own model (you could still refer some k8s type), and only list the fields we need or potential need, so that we could directly UnMarshal raw data to that model we need ( don't need to manually hack abc field to be empty), it will have better performance. Also, that will be self explain what data we collect from customer(easy to communicate to new client), easy to maintain(if we need to collect more data/field, just add new field in that model, is there any challenge blocking you from that option

Reply:
The initial approach only sanitized potentially sensitive data that we very obviously did not need. Peter requested that I do a more in-depth pass to remove all data that we did not need, and this is the data that I determined we do not need. I could potentially create our own model for this, but that puts the burden on us to maintain it and ensure that it is in sync with what we're using on the backend (as we are using the kube object definitions there as well).

1. I agree, I could put a comment along the lines of "we don't use these", but I figured that was sort of implicit.
2. If it's removed and our code isn't updated, no changes/errors occur. If it's removed from the library when we update the version of the dependency, then it wont compile until the reference is removed.

I'll take a look at copying the k8s objects and seeing if that's something we want to maintain.

Reply:
So looking into this, it is definitely do-able, but I personally do not think it is something that we want to maintain.

We would effectively be recreating somewhere between 40-60% of the object definitions from [this file](https://github.com/kubernetes/kubernetes/blob/v1.24.2/pkg/apis/core/types.go).

In my experience, recreating complex contracts with 3rd parties becomes _very_ painful and time intensive as time goes on, and is not something I personally view as worth the hassle of maintaining. If this is something that the team thinks is worth doing, then I am open to pursuing it, but I would strongly advocate against it.

Reply:
If we're blindly bumping major versions we have a more serious problem imo 😄

To your point though, a non-backwards-compatible change would accompany a new kubernetes release and we would notice that issue rather quickly in both the agent & denormalizer when testing it. At the moment we are using a relatively old version of the kube models internally.

I could also add a unit test to make sure we're within some percentage of the original file size.

Reply:
Just want to better understand what's the `pain` you mentioned above. If k8s makes backward compatible change to the field we care, we don't need to make any change to our own model.. If it is non-backward compatible change, I assume we just need to change the json name in our own model(small effort). BUT, we have to change that anyway in our backend system anyway even we use k8s model when we extracting data and put in measure model

Assert the file size is good than nothing! I still hope one day we could reach a point that we assert the data we collect.

Reply:
I preferred to the approach to explicitly keep the data we care than filtering out data we don't need. There could be in the future "extended" fields from those responses via k8s sdk. We may not notice the change, but data ingestion file size could be blown up.